samples/sample_decode/src/pipeline_decode.cpp

```diff
@@ -402,6 +402,39 @@ mfxStatus CDecodingPipeline::Init(sInputParams *pParams)
     sts = m_mfxSession.QueryIMPL(&m_impl); // get actual library implementation
     MSDK_CHECK_STATUS(sts, "m_mfxSession.QueryIMPL failed");
 
+    m_bVppIsUsed = IsVppRequired(pParams);
+    if (m_bVppIsUsed)
+        m_bDecOutSysmem = pParams->bUseHWLib ? false : true;
+    else
+        m_bDecOutSysmem = pParams->memType == SYSTEM_MEMORY;
+
+    m_eWorkMode = pParams->mode;
+
+    m_monitorType = pParams->monitorType;
+    // create device and allocator
+#if defined(LIBVA_SUPPORT)
+    m_libvaBackend = pParams->libvaBackend;
+#endif // defined(MFX_LIBVA_SUPPORT)
```

Comment:
Why do we need this here? I think it should remain where it was. Or I'm missing somethig?

Reply:
@dmitryermilov, we use [m_eWorkMode](https://github.com/Intel-Media-SDK/MediaSDK/blob/406334ff98465e8c5dd8be42b0934c8aac7c772e/samples/sample_decode/src/pipeline_decode.cpp#L1034), [m_monitorType ](https://github.com/Intel-Media-SDK/MediaSDK/blob/406334ff98465e8c5dd8be42b0934c8aac7c772e/samples/sample_decode/src/pipeline_decode.cpp#L1066), [m_libvaBackend](https://github.com/Intel-Media-SDK/MediaSDK/blob/406334ff98465e8c5dd8be42b0934c8aac7c772e/samples/sample_decode/src/pipeline_decode.cpp#L1070)  in CreateHWDevice(). And m_bDecOutSysmem will be needed for this condition `if(m_memType != SYSTEM_MEMORY || !m_bDecOutSysmem)` on windows.

Reply:
Okay.

You know, even code seems to be correct in terms of logic/language mistakes, I somehow don't like how it looks. There are "low level" details which hide initialization steps logic. It's generally a global problem of our samples. Two proposals for current PR:
- please move code at lines 405-413  before GetImpl()
- then do something like :
```
bool isDeviceRequired = false;
mfxHandleType hdl_t = unkown;
#if Windows
  isDeviceRequired = m_memType != SYSTEM_MEMORY || !m_bDecOutSysmem;
  hdl_t = D3D9 or D3D11;
#elseif Linux
  isDeviceRequired = true;
  hdl_t = VA_DISPLAY;
#endif
if (isDeviceRequired) 
{
   CreateHWDevice();
   mfxHDL hdl = NULL;
   m_hwdev->GetHandle(hdl_t, &hdl);
   m_mfxSession.SetHandle(hdl_t, hdl);
}
```

Reply:
@dmitryermilov, done, but the `hdl_t` is not initialized as `unknown` when declared, because it is not provided in `mfxHandleType`.